### ISSUE TITLE
Please *do* generate man pages

### ISSUE THREAD

**Issue by benknoble** [cli-guidelines/cli-guidelines#57]
Not a PR of improvements, but some ideas about why I think this a bad idea (post is mostly un-edited; interested in discussing the issue more than changing minds, but I think it worth pointing out that if you *can* do man-pages, do so! And more systems should make doing so easy.)

https://benknoble.github.io/blog/2020/12/07/man-pages-please/

**Comment by bfirsh** [comment]
That's a really thoughtful blog post -- thank you.

This rule came from me I think, and this is the sentiment I intended (though this was a collaboration so maybe other authors have other opinions): https://news.ycombinator.com/item?id=25317960

I don't want to stop people from writing man pages, but the priority should be directed toward help texts and web pages because more people read them.

Maybe it would benefit from being less strongly worded.

**Comment by aanand** [comment]
Thanks for the post, @benknoble! I’ll offer my opinions, as another author of the document. The high-level summary is that they're almost exactly the same as @bfirsh's.

I don't think anyone here disagrees with the assertion that having comprehensive documentation available in the terminal is a good thing, whatever you type to access it. There are obvious benefits versus web-based documentation: it's much quicker than switching to your browser and loading up a web page or typing into a search engine, and it works offline.

If that documentation is accessible via `man mycmd`, all the better - plenty of people (myself included) look for that. However, it's my suspicion that they (we) constitute a minority of the users of any given command. If `man mycmd` is the only way to find those docs, then only a minority of users (and precisely nobody on Windows) will ever find them. Therefore, it's critical that those docs be accessible via `mycmd help`, `mycmd subcmd help`, etc. (The first tool I ever saw do this was `git`, which is endlessly impressive in how it manages to do some aspects of CLI design so well and others so badly.)

However, zooming out for a moment, what's undeniable is that web-based docs are _much_ more inclusive. Just as not everyone knows what `man` is, not everyone is comfortable navigating documents with a terminal-based pager, or knows how to search in one, or even how to page up and down. What's more, people need to be able to search your docs online, and link their peers and colleagues to specific parts. (And speaking of links, online docs can actually be interlinked, which is a massive usability boost. Unless man pages can do that too and I've missed a trick my whole career?)

To sum up, then:

1. Web-based documentation is top priority.
2. Terminal-based documentation is good.
3. Having your terminal-based documentation accessible via `man mycmd` is good, but having it accessible via a `help` subcommand is much more important. Obviously, these aren't mutually exclusive: do both.

(I've heard a lot of people say "why not generate one from the other?" I think that if you can do that, that's great, but I don't want to constrain anyone's choice of tooling.)

Given all of this, I think "Don't bother with man pages" could do with rewording. The last thing I want to do is actively discourage anyone from making docs available in the terminal - I just want to make sure people prioritise web-based docs.

**Comment by benknoble** [comment]
@bfirsh 

> That's a really thoughtful blog post -- thank you.

More thanks than I deserve for a rant, heh. I read through most of the HN comments, and I think I largely agreed with some sentiments. I saw some of my own arguments, too. (The `apropros`/`man -k` was one I hadn't considered, but it's a good one.)

As for the sentiment… I wonder if `--help` should stay briefer, with a pointer to `--help-verbose` or `--man` or what-have-you for more details à la man-pages? If you absolutely can't have man-pages, that is—many have pointed out that there is lots of tooling for having them (and, fairly enough, it doesn't have to be `man` or even `*roff`—pod, info, texdoc, the list goes on of tool-systems and their specialized documentations. The point is, have some.).

I think a better sentiment is going to be "consider your audience"—it may not be true for all programs that the audience reads less man-pages! Adjust your priorities accordingly.

---

@aanand 

> If that documentation is accessible via `man mycmd`, all the better […]

Yeah, this is fair. It doesn't have to be `man`. But it better be something in the terminal, and `--help` really isn't my personal favorite. There are plenty of standard documentation systems; if you're part of a new ecosystem, maybe it's time to build one. Again, the brevity of `help`-like commands is actually appreciated.

>  not everyone is comfortable navigating documents with a terminal-based pager, or knows how to search in one, or even how to page up and down

Arguable that a CLI user *should* know, but I see your point that the web is more _familiar_ and _navigable_ to the novice.

I, and most others, are not saying "no web pages." I am saying no to "no man pages." It's *not* inclusive to eliminate a form of documentation that is accessible to those who want it. (I don't consider some web documentation as accessible as others, and this is without discussing issues of internet access. HN covered that fairly well.)

> What's more, people need to be able to search your docs online, and link their peers and colleagues to specific parts

Ironically, `git`'s HTML-formatted man-pages provide exactly this…

> And speaking of links, online docs can actually be interlinked, which is a massive usability boost. Unless man pages can do that too and I've missed a trick my whole career?

Covered in HN, but to re-iterate: you can get `cat(1)` into a man-page and most viewers know how to link that (I'm not counting `less`). Some make it an HTML link, a PDF link, or (in vim) I can press `K` and automatically link. See also the tail end of most man-pages (the section called `SEE ALSO`).

Example from `cat(1)`:

```
SEE ALSO
     head(1), more(1), pr(1), sh(1), tail(1), vis(1), zcat(1), setbuf(3)
```

> (I've heard a lot of people say "why not generate one from the other?" I think that if you can do that, that's great, but I don't want to constrain anyone's choice of tooling.)

Indeed, one of my arguments, too. If the program doesn't exist for your ecosystem, maybe it's time to write it…

> Given all of this, I think "Don't bother with man pages" could do with rewording. The last thing I want to do is actively discourage anyone from making docs available in the terminal - I just want to make sure people prioritise web-based docs.

Again, priorities differ depending on audience. I understand CLIG will have some prescriptive parts, but maybe prescribing making an informed choice is a better idea here?
